// Copyright Offene Werkstatt WÃ¤denswil
// SPDX-License-Identifier: MIT

syntax = "proto3";

package maco.factory;

// RPC service for factory hardware testing.
// Used during bring-up and QA to verify individual hardware subsystems
// (LEDs, display, etc.) without running the full application stack.
service FactoryTestService {
  // Set all LEDs to a single color.
  rpc LedSetAll(LedColorRequest) returns (TestResponse);

  // Set a single LED pixel.
  rpc LedSetPixel(LedPixelRequest) returns (TestResponse);

  // Clear all LEDs (turn off).
  rpc LedClear(Empty) returns (TestResponse);

  // Set display backlight brightness.
  rpc DisplaySetBrightness(BrightnessRequest) returns (TestResponse);

  // Fill entire display with a solid color.
  rpc DisplayFillColor(DisplayColorRequest) returns (TestResponse);

  // Show vertical color bar test pattern (R/G/B/W/C/M/Y).
  rpc DisplayColorBars(Empty) returns (TestResponse);

  // Play a buzzer tone at a given frequency and duration.
  rpc BuzzerBeep(BuzzerBeepRequest) returns (TestResponse);

  // Stop buzzer immediately.
  rpc BuzzerStop(Empty) returns (TestResponse);

  // Read capacitive touch button state.
  // Long-polls: blocks until a touch is detected or timeout_ms expires.
  rpc TouchRead(TouchReadRequest) returns (TouchReadResponse);
}

message Empty {}

message TestResponse {
  bool success = 1;
  string message = 2;
}

message LedColorRequest {
  uint32 r = 1;
  uint32 g = 2;
  uint32 b = 3;
  uint32 w = 4;
}

message LedPixelRequest {
  uint32 index = 1;
  uint32 r = 2;
  uint32 g = 3;
  uint32 b = 4;
  uint32 w = 5;
}

message BrightnessRequest {
  uint32 brightness = 1;  // 0-255
}

message DisplayColorRequest {
  uint32 r = 1;
  uint32 g = 2;
  uint32 b = 3;
}

message BuzzerBeepRequest {
  uint32 frequency_hz = 1;
  uint32 duration_ms = 2;
}

message TouchReadRequest {
  uint32 timeout_ms = 1;  // Max wait time in ms (0 = single immediate read)
}

message TouchReadResponse {
  uint32 raw_bitmask = 1;
  bool button_up = 2;
  bool button_down = 3;
  bool button_ok = 4;
  bool button_cancel = 5;
}
